server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java

```diff
@@ -796,6 +799,9 @@ public static Query dateRangeQuery(
                     if (includeLower == false) {
                         ++l;
                     }
+                    if (fieldName.equals(DataStream.TIMESTAMP_FIELD_NAME)) {
```

Comment:
This is the core of the change: as we parse the lower bound of a time range, if it is against the `@timestamp` field, we save the parsed value in the context for later retrieval. This is good so we don't reparse dates multiple times which would be expensive.